python/ray/data/_internal/util.py

```diff
@@ -51,3 +58,94 @@ def _check_pyarrow_version():
             )
         else:
             _VERSION_VALIDATED = True
+
+
+def _autodetect_parallelism(
+    parallelism: int,
+    cur_pg: Optional["PlacementGroup"],
+    reader: Optional["Reader"] = None,
+    avail_cpus: Optional[int] = None,
+) -> (int, int):
+    """Returns parallelism to use and the min safe parallelism to avoid OOMs.
+
+    This detects parallelism based on three heuristic functions:
```

Comment:
Shouldn't ctx.min_parallelism also be included in this list?

Also, I found the heuristic description a little bit hard to follow because I wasn't really sure about priority between the different rules. Might help to also say which cases we cannot avoid breaking a rule. Also to say which ones lower-bound vs upper-bound the parallelism.

Reply:
I updated this to include the min_parallelism and state the precedence in order.